# ufBGCtoolbox
ufBGCtoolbox consists of three modules designed for the mining of BGC domains 
and classes in metagenomic data.  
1. bgc_dom_annot: fast identification of BGC protein domains.  
2. bgc_dom_shannon: BGC domain-specific diversity estimation.  
3. bgc_model_class: BGC class relative count predictions.  

## Install

ufBGCtoolbox consists of four dokcer images: 
1. epereira/ufBGCtolbox:bgc_dom_annot 
2. epereira/ufBGCtolbox:bgc_dom_shannon 
3. epereira/ufBGCtolbox:bgc_dom_merged_shannon 
4. epereira/ufBGCtolbox:bgc_class_models

Before running ufBGCtoolbox it is necessary to install [docker](https://www.docker.com/).

Then just clone the github repository:
```
git clone git@github.com:pereiramemo/ufBGCtoolbox.git
```

All four images are in [dockerhub](https://hub.docker.com/). These will be downloaded automatically the first time you run the scripts.

## Documentation

### 1. bgc_dom_annot
This first module runs uproc using a BGC domain profile database. It takes as an input metagenomic unassembled data and outputs an abundance BGC domain profile table.

### 2. bgc_dom_shannon
bgc_dom_shannon has two different modes: sample and merge. The sample mode takes as an input metagenomic unassembled data and generates a targeted assembly of the domains. Subsequently, it clusters the assembled sequences and places these on pre-computed reference trees and computes the Shannon diversity index. The merge mode integrates the sample mode results and computes the diversity based on rarefied subsamples.

### 3. bgc_class_models
This module is based on the [bgcpred](https://git@github.com:pereiramemo/bgcpred.git) R package, which includes a library of BGC class abundance models. Based on the domain profile generated by bgc_dom_annot, this module computes the BGC class abundance profile.


## ufBGCtoolbox example workflow 

In this example we will analize three different simulated metagenimc samples: sim_meta_oms-1, sim_meta_oms-2 and sim_meta_oms-3.

### bgc_dom_annot

See help
```
sudo ./run_bgc_dom_annot.bash . . --help
```
Run
```
for i in $( seq -s" " 1  3); do

  sudo ./run_bgc_dom_annot.bash \
  example/sim_meta_oms-"${i}"_redu_r1.fasta.gz \
  example/sim_meta_oms-"${i}"_redu_r2.fasta.gz \
  example/out_dom_annot"${i}" \
  --intype dna \
  --nslots 2
  
done

```

### ufBGCtoolbox: bgc_dom_shannon sample

See help
```
sudo ./run_bgc_dom_shannon.bash sample . . . --help

```

With Docker, all input files must be in the same directory

```
for i in $( seq -s" " 1  3); do
  sudo mv example/out_dom_annot"${i}"/pe_bgc_dom.gz example/pe_bgc_dom"${i}".gz
done
  
```
Run bgc_dom_shannon in sample mode
```
for i in $( seq -s" " 1  3); do

  sudo ./run_bgc_dom_shannon.bash sample \
  example/pe_bgc_dom"${i}".gz \
  example/sim_meta_oms-"${i}"_redu_r1.fasta.gz \
  example/sim_meta_oms-"${i}"_redu_r2.fasta.gz \
  example/out_dom_shannon"${i}" \
  --blast t \
  --identity 0.5 \
  --plot_tree t \
  --only_rep t \
  --coverage t \
  --nslots 4 \
  --verbose t \
  --font_size 2 \
  --domains PKS_KS,Condensation
  
done
```

Estimated diversities:

Sample | PKS_KS | Condensation
---|---|---
sim_meta_oms-1 | 3.299 | 4.672
sim_meta_oms-2 | 2.435 | 3.928
sim_meta_oms-3 | 2.408 | 3.712

To visualize the sequence placements the Condensation_placements_tree.pdf and PKS_KS_placements_tree.pdf images are generated:

Placed Condensation sequences from sample sim_meta_oms-1
![tree PKS_KS](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/Condensation_placements.png)


Placed PKS_KS sequences from sample sim_meta_oms-1
![tree PKS_KS](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/PKS_KS_placements.png)

### ufBGCtoolbox: bgc_dom_shannon merge

See help
```
sudo ./run_bgc_dom_shannon.bash merge . . --help

```

Run bgc_dom_shannon in merge mode for Condensation
```
sudo ./run_bgc_dom_shannon.bash merge \
example/out_dom_shannon1,example/out_dom_shannon2,example/out_dom_shannon3 \
example/out_dom_merged_shannon_Condensation \
--domain Condensation \
--num_iter 50 \
--sample_increment 20 \
--plot_rare_curve t \
--plot_tree t \
--only_rep t \
--verbose t
```
This will output the rarefaction and tree plots to compare the domain diversity and composition between the three samples.

Placed Condensation sequences
![tree PKS_KS](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/Condensation_merged_placements.png)

Rarefied Condensation diversity

![rare Condensation](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/Condensation_rare_div_est.png)


Run bgc_dom_shannon in merge mode for PKS_KS
```
sudo ./run_bgc_dom_shannon.bash merge \
example/out_dom_shannon1,example/out_dom_shannon2,example/out_dom_shannon3 \
example/out_dom_merged_shannon_PKS_KS \
--domain PKS_KS \
--num_iter 50 \
--sample_increment 20 \
--plot_rare_curve t \
--plot_tree t \
--only_rep t \
--verbose t

```
Placed PKS_KS sequences
![tree PKS_KS](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/PKS_KS_merged_placements.png)



Rarefied PKS_KS diversity
![rare PKS_KS](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/PKS_KS_rare_div_est.png)

### ufBGCtoolbox: bgc_model_class	

See help
```
sudo ./run_bgc_class_models.bash . . --help
```

Run 
```
for i in $( seq -s" " 1  3); do
sudo ./run_bgc_class_models.bash \
  example/out_dom_annot"${i}"/counts.tbl \
  example/out_class_models"${i}"
done

```

Predicted abundances :

sim_meta_oms-1

![barplot1](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/bgc_class_pred1.png)

sim_meta_oms-2

![barplot2](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/bgc_class_pred2.png)

sim_meta_oms-3

![barplot3](https://github.com/pereiramemo/ufBGCtoolbox/blob/master/example/bgc_class_pred3.png)